peer_client.go

```diff
@@ -144,19 +138,25 @@ func (c *PeerClient) connect(ctx context.Context) (err error) {
 		}
 
 		var err error
+		c.clientMutex.Lock()
 		c.conn, err = grpc.Dial(c.conf.Info.GRPCAddress, opts...)
 		if err != nil {
+			c.clientMutex.Unlock()
 			return c.setLastErr(&PeerErr{err: errors.Wrapf(err, "failed to dial peer %s", c.conf.Info.GRPCAddress)})
 		}
 		c.client = NewPeersV1Client(c.conn)
```

Comment:
I didn't understand why every API attempts to create a new connection. Why not create one in `NewPeerClient`?

Reply:
We were new to GRPC and were used to having to manage the connection reconnect.  It's very strange that when using GRPC the user has to pass in the a new connection to make a GRPC client, but then the GRPC client is expected to restart the connection. Because we passed in the connection, we assumed (incorrectly) that we also had to manage the connect reconnect. Also, if I recall correctly, we didn't want to leave open connections to each peer for even if they were not used. But now we know, that is how GRPC is supposed to work.

I removed all of this crap when I removed GRPC in #206. For those still on V2, I never saw a need to refactor all this, if you plan on using v2 for a while and have an interest, you can refactor this if you wish!

Reply:
When you say "for those still on v2", what do you mean? Isn't everyone on v2?

In this pr I wanted to fix the flaky test but I realized that unless I make major changes (ie remove all the mutex logic), it's pretty much impossible. So I'm not sure how to proceed. Are there plans to release v3 soon? I haven't seen your PR in detail

Reply:
"soon" is at least 3 weeks away as I'm traveling for the next 2 weeks. I leave it up to you if you wish to refactor the peer client to remove the mutexes and rely on GRPC for reconnect.